File: search_cpp/min_max_stats.cpp
#include "min_max_stats.hpp"

#include <algorithm>
#include <limits>
#include <sstream>
#include <stdexcept>

namespace rainbow::search {

MinMaxStats::MinMaxStats(
    const std::vector<double>& known_bounds,
    const bool soft_update,
    const double min_max_epsilon)
    : soft_update_(soft_update),
      min_max_epsilon_(min_max_epsilon),
      max_(known_bounds.size() >= 2 ? known_bounds[1] : -std::numeric_limits<double>::infinity()),
      min_(known_bounds.size() >= 2 ? known_bounds[0] : std::numeric_limits<double>::infinity()) {
    if (known_bounds.size() == 1 || known_bounds.size() > 2) {
        throw std::invalid_argument("known_bounds must contain either 0 or 2 values.");
    }
}

void MinMaxStats::update(const double value) {
    max_ = std::max(max_, value);
    min_ = std::min(min_, value);
}

double MinMaxStats::normalize(const double value) const {
    if (max_ > min_) {
        double denom = max_ - min_;
        if (soft_update_) {
            denom = std::max(denom, min_max_epsilon_);
        }
        return (value - min_) / denom;
    }
    return value;
}

double MinMaxStats::min() const {
    return min_;
}

double MinMaxStats::max() const {
    return max_;
}

bool MinMaxStats::soft_update() const {
    return soft_update_;
}

double MinMaxStats::min_max_epsilon() const {
    return min_max_epsilon_;
}

std::string MinMaxStats::repr() const {
    std::ostringstream out;
    out << "min: " << min_ << ", max: " << max_;
    return out.str();
}

}  // namespace rainbow::search

File: search/search_py/min_max_stats.py
import collections
from typing import Optional, List

MAXIMUM_FLOAT_VALUE = float("inf")

# TODO: EFFICIENT ZERO SOFT MINMAX STATS


class MinMaxStats(object):
    def __init__(
        self,
        known_bounds: Optional[List[float]],
        soft_update: bool = False,
        min_max_epsilon: float = 0.01,
    ):
        self.soft_update = soft_update
        self.min_max_epsilon = min_max_epsilon
        self.max = known_bounds[1] if known_bounds else -MAXIMUM_FLOAT_VALUE
        self.min = known_bounds[0] if known_bounds else MAXIMUM_FLOAT_VALUE

    def update(self, value: float):
        self.max = max(self.max, value)
        self.min = min(self.min, value)

    def normalize(self, value):
        epsilon = self.min_max_epsilon if self.soft_update else 1e-8

        # DeepMind mctx logic: If stats are uninitialized (max < min),
        # pretend max = min = value. This ensures the numerator is 0.0
        # and natural evaluation forces the uninitialized Q-value to 0.0
        # rather than returning un-normalized values or crashing.
        is_uninitialized = self.max < self.min
        maximum = value if is_uninitialized else self.max
        minimum = value if is_uninitialized else self.min

        diff = maximum - minimum

        # Denominator calculation (works for both Tensors and standard floats)
        import torch

        if isinstance(diff, torch.Tensor):
            denom = torch.clamp(diff, min=epsilon)
        else:
            denom = max(diff, epsilon)

        return (value - minimum) / denom

    def __repr__(self):
        return f"min: {self.min}, max: {self.max}"
